applovin_max/ios/Classes/AppLovinMAX.m

```diff
@@ -1601,6 +1578,26 @@ + (CGSize)adViewSizeForAdFormat:(MAAdFormat *)adFormat
              @"waterfall": [self createAdWaterfallInfo: ad.waterfall]};
 }
 
+- (NSDictionary<NSString *, id> *)adLoadFailedInfoForAd:(NSString *)adUnitIdentifier withError:(MAError *)error
+{
+    return ( error ) ?
+    @{@"adUnitId": adUnitIdentifier,
+      @"code" : @(error.code),
+      @"message" : error.message,
+      @"waterfall": [self createAdWaterfallInfo: error.waterfall]}
+    :
+    @{@"adUnitId": adUnitIdentifier,
+      @"code" : @(MAErrorCodeUnspecified)};
+}
+
+- (NSDictionary<NSString *, id> *)adDisplayFailedInfoForAd:(MAAd *)ad withError:(MAError *)error
+{
+    NSMutableDictionary *body = [NSMutableDictionary dictionaryWithCapacity: 2];
+    body[@"ad"] = [self adInfoForAd: ad];
+    body[@"error"] = [self adLoadFailedInfoForAd: ad.adUnitIdentifier withError: error];
```

Comment:
Does `adInfoForAd:` or `adLoadFailedInfoForAdUnitIdentifier:error:` return `nil`? If not, you can just return the dictionary `@{}` literal like

```
return @{
    @"ad" : [self adInfoForAd: ad],
    @"error" : [self adLoadFailedInfoForAdUnitIdentifier: ad.adUnitIdentifier error: error]
};
```